In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
from torch.distributions import Normal
import torch.optim as optim
from RL_brain import QLearningTable
# 检查GPU可用性
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# 已有数据（保持不变）
seed_value = 1234
np.random.seed(seed_value)
UNIT = 40
MAZE_H = 8
MAZE_W = 8
utilization_ratios_device = 0.1
utilization_ratios_server = 0.1
car_flop = 1.3 * 10**12
car_power = 20
UAV_flop = 0.641 * 10**12
# UAV_power = 20##30
UAV_power = 30
e_flop = 330 * 10**12
e_power = 450
d_fai = 5*10**-29
trans_v_up = 100*1024*1024/4 #553081138.4484484
trans_v_dn = 20*1024*1024/4
p_cm = 0.1
# nums_data = np.array([5, 3, 4, 7, 9])  # 客户端本地数据量
partition_point = [0, 1, 2, 3, 4, 5, 6]
num_img_UAV = 3
num_img_car = 1

device_load = [0.3468e9, 0.3519e9, 2.3408e9, 2.3409e9, 5.3791e9, 9.6951e9, 12.077e9]
server_load = [11.7321e9, 11.727e9, 9.7381e9, 9.738e9, 6.6998e9, 2.3838e9, 0.0019e9]
exchanged_data = [2359296, 2359296, 2359296, 2359296, 1179628, 589824, 294912]
privacy_leak = [0.96122, 0.608901, 0.57954889, 0.593044, 0.535525, 0.007155, 0.054303]

# 转换为NumPy数组
np_partition = np.array(partition_point)
np_device = np.array(device_load)
np_server = np.array(server_load)
np_exchanged = np.array(exchanged_data)
np_privacy = np.array(privacy_leak)



Using device: cuda


In [2]:
# index_temp = 1
# avg_t = 0
# r_i_u = 0
# for temp_2 in range(5):
#     b_i = 20e6  # 信道带宽 (Hz)
#     P_i_cm = 0.2  # 车载设备传输功率 (W)
    
#     r = 1.5  # 路径损耗指数
    
#     d_i = np.array([1000.0, 800.0, 600.0, 400.0, 200.0])  # 到边缘服务器的距离 (m)
#     # 计算信道增益
#     h_i = 1 / (d_i**r)
#     h_0 = 0.5
#     noise_density_dBm = -174  # 噪声功率谱密度 (dBm/Hz)
#     # 将噪声功率谱密度从 dBm/Hz 转换为 W/Hz
#     noise_density_W = 10**(noise_density_dBm / 10) * 1e-3  # 转换为 W/Hz
    
#     # 计算噪声功率
#     noise_power = noise_density_W * b_i
    
#     D_i = exchanged_data[6]#####
#     D_i_bits_actual = (D_i / 2) * 2*2  # 每数据点占用 16 位，单位为比特
#     r_i_u += b_i * np.log2(1 + (P_i_cm * h_i[temp_2]) / noise_power)###计算平均值
    
#     t_i_u = 7*D_i_bits_actual / r_i_u####时间
#     print(t_i_u)
#     e_i_cm = P_i_cm * t_i_u###能耗

# print(r_i_u/5)

In [3]:
def cost_cal(num_data, v_flop, device_power, partition_index):
    partial_device = np_device[partition_index]
    device_time = partial_device * num_data / (v_flop *utilization_ratios_device)

    partial_server = np_server[partition_index]
    server_time = partial_server * num_data / ( e_flop* utilization_ratios_server + 1e-8)
    # print(f"device_time is {device_time}, server_time is {server_time}, cal_time is {device_time+server_time}")

    feature = np_exchanged[partition_index]
    trans_t_up = feature / trans_v_up * num_data
    # print(f"device_time is {device_time}, server_time is {server_time}, cal_time is {device_time+server_time},trans_t_up is {trans_t_up}")
    energy_cal = ((partial_device * device_power) / v_flop + (
            partial_server * e_power * utilization_ratios_server) / e_flop) * num_data
    energy_trans = num_data * p_cm * trans_t_up
    energy = energy_cal + energy_trans
    # print(f"energy cal is{energy_cal}, trans is {energy_trans}")
    landa_trans = 1
    time_all = device_time + server_time + landa_trans * trans_t_up
    return time_all, energy

In [4]:

for partition_num in range(len(partition_point)):
    time_UAV, energy_UAV = cost_cal(3, UAV_flop,30 , partition_num)
    # time_car, energy_car = cost_cal(1, car_flop, partition_num)
            # total_time += time
    reward = -(time_UAV)*0.4 - (energy_UAV)*0.3 - 0.3*np_privacy[partition_num]
    # print(f"partition is {partition_num+1}:time is {time_car+time_UAV}, energy is {energy_car + energy_UAV}, reward is {reward}")
    print(f" reward is {reward}")
    
print(f"______________")
for partition_num in range(len(partition_point)):
    # time_UAV, energy_UAV = cost_cal(3, UAV_flop, partition_num)
    time_car, energy_car = cost_cal(1, car_flop, 20, partition_num)
            # total_time += time
    reward = -(time_car)*0.4 - (energy_car)*0.3 - 0.3*np_privacy[partition_num]
    # print(f"partition is {partition_num+1}:time is {time_car+time_UAV}, energy is {energy_car + energy_UAV}, reward is {reward}")
    print(f" reward is {reward}")

 reward is -0.4436326264607857
 reward is -0.3382464115090058
 reward is -0.4501338789805701
 reward is -0.454188480314849
 reward is -0.5551497908531352
 reward is -0.6254742198482486
 reward is -0.7676227740355978
______________
 reward is -0.33035584912587407
 reward is -0.22469910944055943
 reward is -0.23108723524009325
 reward is -0.23513653216783215
 reward is -0.2217401566679699
 reward is -0.08652560594405594
 reward is -0.11402850075757574


### 第一部分 reward 数据
| 序号 | reward 值 |
| ---- | ---- |
| 1 | -0.44254016276223773 |
| 2 | -0.3371378821678322 |
| 3 | -0.4427600640279721 |
| 4 | -0.4468143503496504 |
| 5 | -0.5382049480594235 |
| 6 | -0.5949334332167834 |
| 7 | -0.7295787022727273 |

### 第二部分 reward 数据
| 序号 | reward 值 |
| ---- | ---- |
| 1 | -0.33035584912587407 |
| 2 | -0.22469910944055943 |
| 3 | -0.23108723524009325 |
| 4 | -0.23513653216783215 |
| 5 | -0.2217401566679699 |
| 6 | -0.08652560594405594 |
| 7 | -0.11402850075757574 |

In [5]:
class environment(object):
    def __init__(self, F_step_size=1, F_max=6, F_min=0):
        super(environment, self).__init__()
        self.server_power_step_size = F_step_size
        self.server_power_max = F_max
        self.server_power_min = F_min
        #self.num_user = device_index_1#####第几号设备

        self.action_space = ['l', 'r', 'no']
        self.n_actions = len(self.action_space)
        # self.title('cut the net')

        # create grids
        # for c in range(0, MAZE_W * UNIT, UNIT):####宽width
        #     x0, y0, x1, y1 = c, 0, c, MAZE_H * UNIT
        #     self.canvas.create_line(x0, y0, x1, y1)#######画垂直线，起点坐标 (c, 0) 和终点坐标（c，w）
        # for r in range(0, UNIT, UNIT):
        #     x0, y0, x1, y1 = 0, r, UNIT, r#####画横线，起点坐标 (0, r) 和终点坐标（h，r）
        #     self.canvas.create_line(x0, y0, x1, y1)

        # create origin
#         origin = np.array([20, 20])
#         self.rect = self.canvas.create_rectangle(
#             origin[0] - 15, origin[1] - 15,
#             origin[0] + 15, origin[1] + 15,
#             fill='red')

#         # pack all
#         self.canvas.pack()
    # def reset(self):#########重新开始
    #     self.update()
    #     time.sleep(0.5)
    #     self.canvas.delete(self.rect)
    #     origin = np.array([20, 20])
    #     self.rect = self.canvas.create_rectangle(
    #         origin[0] - 15, origin[1] - 15,
    #         origin[0] + 15, origin[1] + 15,
    #         fill='red')
    #     return self.canvas.coords(self.rect)

    def step(self, state, action, device_index):
        
        state_next = 0
        # s = self.canvas.coords(self.rect)
        # base_action = np.array([0, 0])
        #print(action)

        if action == 1:   # 右边
            if state < 6:
                
                state_next = state+1
                   # record_action = 5
        elif action == 0:   # 左
            if state>0:
                state_next = state -1
        elif action == 2:###no move
            state_next = state
        #print(base_action[0])
        setpoint = state_next
        
            #print(0)
        if device_index == 0:
            num_data = 3  # 假定每次选择相同数量的数据
            v_flop = 0.641 * 10**12
            device_power = 30
        else:
            num_data = 1
            v_flop = 1.3 * 10**12
            device_power = 20
        
        time, energy = cost_cal(num_data, v_flop, device_power, setpoint)
        # time_car, energy_car = cost_cal(num_img_car, e_power, action_car)
            # total_time += time
        reward = -time*0.4 - (energy)*0.3 - 0.3*np_privacy[setpoint]
        # reward = -(time_car+time_UAV)*0.4 - (energy_car + energy_UAV)*0.1 + 0.5*np_privacy[partition_num]
    

        done = False#####不停，后期加停止命令
        
        return state_next, reward, done

In [6]:
# # index_1 = 0
# # num_data_temp = nums_data[index_1]###选着对应设备的对应训练个数nums_data = np.array([54, 30, 40, 91, 92])  # 客户端本地数据量
# #         #print(device_index_1)
# # server_power_temp = servers_power[index_1]
# # cost_time_temp , cost_energy_temp = cost_cal(num_data_temp, server_power_temp)
# servers_power_temp = [0.2, 0.2, 0.2, 0.2, 0.2]
# reward = 0
# for setpoint in range(5):
    
#     num_data_temp = nums_data[3]###选着对应设备的对应训练个数nums_data = np.array([54, 30, 40, 91, 92])  # 客户端本地数据量
#             #print(device_index_1)
#     server_power_temp = servers_power_temp[setpoint]
#     cost_time_temp , cost_energy_temp = cost_cal(1, server_power_temp)
#     landa_1 = 1########计算时间的权重
#     landa_2 = landa_1
#     reward += -1*(landa_1*cost_time_temp[4])# + landa_2*cost_energy[setpoint])
 
# print(f"五个设备的推理总时间为{reward*5}")

In [7]:
# import pandas as pd
# def update2():
#     #num_step = 1
#     # global fun_0
#     # fun_0 = 1
#     # action_note = 1
#     #updated_q_table = pd.DataFrame(RL.q_table)
    
#     initial_qtable = pd.DataFrame(columns=[0, 1, 2], dtype=np.float64)
#     qtables = [initial_qtable.copy() for _ in range(2)]
#     # reward_list = [0.0001]*2
#     # action_list = [0]*2
#     # num_imgs = [3, 1]
#     # num_flops = [1.3 * 10**12 *0.4, 1.3 * 10**12]
#     for i in range(15000):####一共就训练十轮
        
#         device_index = i%2
#         #print(device_index)
#         #print(f"Starting training with Qtable{device_index + 1}")
        
#         RL = QLearningTable(actions=list(range(env.n_actions)), q_table=qtables[device_index])
        
#         # trained_qtable = train_with_qtable(rl)
#         # qtables[i] = trained_qtable
#         record_1 = 0
#         observation = 0
#         while True:
#             #device_index_1 = device_index
#             action = RL.choose_action(str(observation), 1)
#                 #print(action)
#                 #print(format(observation))
#             observation_, reward, done = env.step(observation, action, device_index)
#             RL.learn(str(observation), action, reward, str(observation_))
            
#             observation = observation_
#             # if record_1+1 % 2000== 0:
#             #     print(reward)
#                 #print(device_index_1)
#             record_1 += 1    
            
#             # if record_1 % 400 == 0:
#             #     if device_index == 1:
#             #         print(f"action is {action}")
            
#                     # print(f"reward is {all_cost}")
#             if record_1 > 10:
#                 # reward_list[device_index] = abs(reward)
#                 # print(f"device {device_index}轮次 {i}split-point is {observation_},and reward {abs(reward)}")
#                 if device_index ==0:
#                     history_0.append(reward)
#                 else:
#                     history_1.append(reward)
                        
#                 trained_qtable = RL.q_table
#                 qtables[device_index] = trained_qtable
                
                
#                 #print(f"device{device_index+1} is {reward_list[device_index]}")
#                 break
#         ###每五次都要输出一次总延时
        
    
# history_1 = []
# env = environment()
# # print("env.n_actions: {}".format(env.n_actions))
# #RL = QLearningTable(actions=list(range(env.n_actions)))
# history_0 = []

# update2()
# # print(Q_history)

In [8]:
# # 假设 history_0 是你的长列表
# for item in history_1:
#     print(item)  # 每个元素单独一行，自动换行

In [9]:
# # 
# device 0轮次 0split-point is 0,and reward 0.4436326264607857
# device 1轮次 1split-point is 5,and reward 0.08652560594405594
# device 0轮次 2split-point is 1,and reward 0.3382464115090058
# device 1轮次 3split-point is 3,and reward 0.23513653216783215
# device 0轮次 4split-point is 0,and reward 0.4436326264607857
# device 1轮次 5split-point is 5,and reward 0.08652560594405594
# device 0轮次 6split-point is 1,and reward 0.3382464115090058
# device 1轮次 7split-point is 5,and reward 0.08652560594405594
# device 0轮次 8split-point is 1,and reward 0.3382464115090058
# device 1轮次 9split-point is 5,and reward 0.08652560594405594
# device 0轮次 10split-point is 1,and reward 0.3382464115090058
# device 1轮次 11split-point is 5,and reward 0.08652560594405594
# device 0轮次 12split-point is 1,and reward 0.3382464115090058
# device 1轮次 13split-point is 5,and reward 0.08652560594405594
# device 0轮次 14split-point is 1,and reward 0.3382464115090058
# device 1轮次 15split-point is 5,and reward 0.08652560594405594
# device 0轮次 16split-point is 1,and reward 0.3382464115090058
# device 1轮次 17split-point is 5,and reward 0.08652560594405594
# device 0轮次 18split-point is 1,and reward 0.3382464115090058
# device 1轮次 19split-point is 5,and reward 0.08652560594405594
# device 0轮次 20split-point is 1,and reward 0.3382464115090058
# device 1轮次 21split-point is 5,and reward 0.08652560594405594
# device 0轮次 22split-point is 1,and reward 0.3382464115090058
# device 1轮次 23split-point is 5,and reward 0.08652560594405594
# device 0轮次 24split-point is 1,and reward 0.3382464115090058
# device 1轮次 25split-point is 5,and reward 0.08652560594405594
# device 0轮次 26split-point is 1,and reward 0.3382464115090058
# device 1轮次 27split-point is 5,and reward 0.08652560594405594
# device 0轮次 28split-point is 1,and reward 0.3382464115090058
# device 1轮次 29split-point is 5,and reward 0.08652560594405594
# device 0轮次 30split-point is 1,and reward 0.3382464115090058
# device 1轮次 31split-point is 5,and reward 0.08652560594405594
# device 0轮次 32split-point is 1,and reward 0.3382464115090058
# device 1轮次 33split-point is 5,and reward 0.08652560594405594
# device 0轮次 34split-point is 1,and reward 0.3382464115090058
# device 1轮次 35split-point is 5,and reward 0.08652560594405594
# device 0轮次 36split-point is 1,and reward 0.3382464115090058
# device 1轮次 37split-point is 5,and reward 0.08652560594405594
# device 0轮次 38split-point is 1,and reward 0.3382464115090058
# device 1轮次 39split-point is 5,and reward 0.08652560594405594

In [10]:
import pandas as pd
def update():#####这是多步的Qlearning
    #global fun_1
    
    num_step = 3
    gamma = 0.5  # reward_decay=0.5
    # 初始化记录变量

    record_1 = 0
    sum_reward = [0] * num_step
    action_note = False
    initial_qtable = pd.DataFrame(columns=[0, 1, 2], dtype=np.float64)
    qtables = [initial_qtable.copy() for _ in range(2)]
    
    for i in range(1500):####一共就训练十轮
        
        
        device_index = i%2
        #print(device_index)
        #print(f"Starting training with Qtable{device_index + 1}")
        
        RL = QLearningTable(actions=list(range(env.n_actions)), q_table=qtables[device_index])
        
        # trained_qtable = train_with_qtable(rl)
        # qtables[i] = trained_qtable
        record_1 = 0
        observation = 0
        
        while True:
            #device_index_1 = device_index
            observation_0 = observation

            for num in range(num_step):
                # 选择并执行动作
                action = RL.choose_action(str(observation), 1)
                #env.render()
                observation_, reward, done = env.step(observation, action, device_index)
                    # 更新奖励计算
                sum_reward[num] = reward + (gamma * sum_reward[num-1] if num > 0 else 0)
                observation = observation_
    
                    # 学习过程
                #print(f"Current reward: {reward}")
                RL.learn(str(observation_0), action, 1*sum_reward[num], str(observation_))
                record_1 += 1
            if record_1 > 10:
                # reward_list[device_index] = abs(reward)
                # print(f"device {device_index}轮次 {i}split-point is {observation_},and reward {abs(reward)}")
                trained_qtable = RL.q_table
                qtables[device_index] = trained_qtable
                if device_index == 0:
                    multi_history_0.append(reward)
                else:
                    multi_history_1.append(reward)
                if i >=200:
                    num_step = min(3, num_step-2) if num_step >= 3 else num_step
                break
                
#                 if record_1 >1000:
                    
#                     num_step = min(3, num_step-2) if num_step >= 3 else num_step
#                     print(f"device {device_index}轮次 {i}split-point is {observation_},and reward {abs(reward)}")
#                     trained_qtable = RL.q_table
#                     qtables[device_index] = trained_qtable
        
#                     break
            
            
multi_history_0 = []
multi_history_1 = []
env = environment()

update()


device 0轮次 0split-point is 2,and reward 0.4501338789805701
device 1轮次 1split-point is 2,and reward 0.23108723524009325
device 0轮次 2split-point is 2,and reward 0.4501338789805701
device 1轮次 3split-point is 2,and reward 0.23108723524009325
device 0轮次 4split-point is 1,and reward 0.3382464115090058
device 1轮次 5split-point is 3,and reward 0.23513653216783215
device 0轮次 6split-point is 1,and reward 0.3382464115090058
device 1轮次 7split-point is 0,and reward 0.33035584912587407
device 0轮次 8split-point is 1,and reward 0.3382464115090058
device 1轮次 9split-point is 0,and reward 0.33035584912587407
device 0轮次 10split-point is 2,and reward 0.4501338789805701
device 1轮次 11split-point is 3,and reward 0.23513653216783215
device 0轮次 12split-point is 0,and reward 0.4436326264607857
device 1轮次 13split-point is 3,and reward 0.23513653216783215
device 0轮次 14split-point is 1,and reward 0.3382464115090058
device 1轮次 15split-point is 2,and reward 0.23108723524009325
device 0轮次 16split-point is 2,and reward 0

In [15]:
# 假设 history_0 是你的长列表
for item in multi_history_1:
    print(item)  # 每个元素单独一行，自动换行

-0.23108723524009325
-0.23108723524009325
-0.23513653216783215
-0.33035584912587407
-0.33035584912587407
-0.23513653216783215
-0.23513653216783215
-0.23108723524009325
-0.23108723524009325
-0.33035584912587407
-0.33035584912587407
-0.2217401566679699
-0.23513653216783215
-0.23108723524009325
-0.08652560594405594
-0.08652560594405594
-0.23108723524009325
-0.23513653216783215
-0.23108723524009325
-0.22469910944055943
-0.23108723524009325
-0.23513653216783215
-0.33035584912587407
-0.33035584912587407
-0.33035584912587407
-0.33035584912587407
-0.33035584912587407
-0.22469910944055943
-0.23108723524009325
-0.33035584912587407
-0.33035584912587407
-0.23513653216783215
-0.11402850075757574
-0.22469910944055943
-0.08652560594405594
-0.22469910944055943
-0.22469910944055943
-0.23513653216783215
-0.22469910944055943
-0.22469910944055943
-0.22469910944055943
-0.22469910944055943
-0.33035584912587407
-0.2217401566679699
-0.11402850075757574
-0.22469910944055943
-0.08652560594405594
-0.231087235240

In [12]:
# device 0轮次 0split-point is 2,and reward 0.4501338789805701
# device 1轮次 1split-point is 4,and reward 0.2217401566679699
# device 0轮次 2split-point is 4,and reward 0.5551497908531352
# device 1轮次 3split-point is 6,and reward 0.11402850075757574
# device 0轮次 4split-point is 1,and reward 0.3382464115090058
# device 1轮次 5split-point is 5,and reward 0.08652560594405594
# device 0轮次 6split-point is 1,and reward 0.3382464115090058
# device 1轮次 7split-point is 5,and reward 0.08652560594405594
# device 0轮次 8split-point is 1,and reward 0.3382464115090058
# device 1轮次 9split-point is 5,and reward 0.08652560594405594
# device 0轮次 10split-point is 1,and reward 0.3382464115090058
# device 1轮次 11split-point is 5,and reward 0.08652560594405594
# device 0轮次 12split-point is 1,and reward 0.3382464115090058
# device 1轮次 13split-point is 5,and reward 0.08652560594405594
# device 0轮次 14split-point is 1,and reward 0.3382464115090058
# device 1轮次 15split-point is 5,and reward 0.08652560594405594
# device 0轮次 16split-point is 1,and reward 0.3382464115090058
# device 1轮次 17split-point is 5,and reward 0.08652560594405594
# device 0轮次 18split-point is 1,and reward 0.3382464115090058
# device 1轮次 19split-point is 5,and reward 0.08652560594405594
# device 0轮次 20split-point is 1,and reward 0.3382464115090058
# device 1轮次 21split-point is 5,and reward 0.08652560594405594
# device 0轮次 22split-point is 1,and reward 0.3382464115090058
# device 1轮次 23split-point is 5,and reward 0.08652560594405594
# device 0轮次 24split-point is 1,and reward 0.3382464115090058
# device 1轮次 25split-point is 5,and reward 0.08652560594405594
# device 0轮次 26split-point is 1,and reward 0.3382464115090058
# device 1轮次 27split-point is 5,and reward 0.08652560594405594
# device 0轮次 28split-point is 1,and reward 0.3382464115090058
# device 1轮次 29split-point is 5,and reward 0.08652560594405594
# device 0轮次 30split-point is 1,and reward 0.3382464115090058
# device 1轮次 31split-point is 5,and reward 0.08652560594405594
# device 0轮次 32split-point is 1,and reward 0.3382464115090058
# device 1轮次 33split-point is 5,and reward 0.08652560594405594
# device 0轮次 34split-point is 1,and reward 0.3382464115090058
# device 1轮次 35split-point is 5,and reward 0.08652560594405594
# device 0轮次 36split-point is 1,and reward 0.3382464115090058
# device 1轮次 37split-point is 5,and reward 0.08652560594405594
# device 0轮次 38split-point is 1,and reward 0.3382464115090058
# device 1轮次 39split-point is 5,and reward 0.08652560594405594

In [13]:
# %run run_this_DQN.py
Q_history = [0.3979611041451284, 0.5084312459310265, 0.45997046104223294, 0.4155057514597751, 0.459547224099518, 0.4462378191065669, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756]
pro_Q_history = [0.3968158409396685, 0.4617453188070799, 0.37802200520249274, 0.44907572209904656, 0.4253692783193946, 0.373271094649017, 0.35262858285274096, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756]
DQN_history_1 = [0.48700546107656695, 0.4611621587522712, 0.35139747398790977, 0.42886654008095887, 0.3894838051213463, 0.3230214205953048, 0.34388482811635274, 0.3494891499487529, 0.4029318371657479, 0.35885990446950455, 0.32245772397272465, 0.4271238647317376, 0.3223334323483416, 0.3222856011367443, 0.4156821318919513, 0.47961136324408704, 0.47453006798650277, 0.43264464520559187, 0.38520215257316476, 0.322121430158435]
multi_NET_history = [0.4147439057662382, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756, 0.3250607373578756]
step3_Q_history = [0.45939951965947967, 0.49532890232914795, 0.4593907102875944, 0.49418325389430184, 0.4342156428077455, 0.47603676515763577, 0.4580774311498178, 0.46602833710329405, 0.4771788679887338, 0.44704993493031153, 0.4615117340782279, 0.4949447807301173, 0.4906639463985181, 0.5587528338537002, 0.4435902799018353, 0.4995112137125783, 0.5238251378023414, 0.4647846076181629, 0.49834294905305476, 0.5038689740782576]
step5_Q_history = [0.49125958703019246, 0.43426354849160065, 0.4799999312991409, 0.44581651997374083, 0.5041475705226472, 0.4778233877080914, 0.4526234586215993, 0.500528963498544, 0.49725438862855054, 0.4952065676256895, 0.47263491436082666, 0.46274462002856187, 0.46804591958394004, 0.4748613268615772, 0.4879815723694093, 0.4456494800301122, 0.46331682965801035, 0.45936754510158695, 0.5044860959038258, 0.4633104089119022]

In [14]:
x = list(range(len(pro_Q_history)))
y = Q_history
y_pro = pro_Q_history
plt.figure(figsize=(10, 6))
import matplotlib as mpl
# 统一长度，取最短的长度
mpl.rcParams['font.family'] = 'Times New Roman'
mpl.rcParams['font.size'] = 20

# 假设x是各模型对应的epoch列表（这里假设每个历史列表长度相同，实际需根据数据调整）
# 定义函数获取每个列表的最小值及其索引
def get_min_point(history_list):
    min_value = min(history_list)
    min_index = history_list.index(min_value)
    return min_index, min_value

# 假设各历史列表：y, y_pro, step5_Q_history, step3_Q_history, DQN_history_1, multi_NET_history
# 分别获取每个列表的最小点
q_learning_min = get_min_point(y)
dynamic_multi_step_min = get_min_point(y_pro)
step5_q_min = get_min_point(step5_Q_history)
step3_q_min = get_min_point(step3_Q_history)
dqn_min = get_min_point(DQN_history_1)
multi_net_min = get_min_point(multi_NET_history)

# 绘制散点图

plt.scatter(q_learning_min[0], q_learning_min[1], marker='o', color='#fcf1f0', s=200, edgecolors='black', linewidths=0.8, label='Q - learning')
plt.scatter(dynamic_multi_step_min[0], dynamic_multi_step_min[1], marker='s', color='#fcccbc', edgecolors='black', linewidths=0.8, label='dynamic multi - step Q - learning')
plt.scatter(step5_q_min[0], step5_q_min[1], marker='^', color='#bdb5e1', s=200, edgecolors='black', linewidths=0.8, label='5 - step Q - learning')
plt.scatter(step3_q_min[0], step3_q_min[1], marker='d', color='#b0d992', s=200, edgecolors='black', linewidths=0.8, label='3 - step Q - learning')
plt.scatter(dqn_min[0], dqn_min[1], marker='*', color='#f9d580', s=200, edgecolors='black', linewidths=0.8, label='DQN')
plt.scatter(multi_net_min[0], multi_net_min[1], marker='*', color='#99b9e9', s=200, edgecolors='black', linewidths=0.8, label='our algorithm')
# 调整轴标签字体大小
plt.xlabel('Epoch', fontsize=20)
plt.ylabel('Time Consumption', fontsize=20)
plt.grid(True, linestyle='--', alpha=0.5)
# 调整图例字体大小
plt.legend(fontsize=16)
plt.show()

NameError: name 'plt' is not defined

In [ ]:
#plt.savefig('G:/XYB/globecom/RL.png')  
plt.savefig('G:/XYB/globecom/picture/RL.png', dpi=300, bbox_inches='tight')  
plt.show()  

In [ ]:
x = list(range(len(pro_Q_history)))
y = Q_history
y_pro = pro_Q_history
plt.figure(figsize=(20, 16))
# 设置字体为Times New Roman并放大
mpl.rcParams['font.family'] = 'Times New Roman'
mpl.rcParams['font.size'] = 20  # 可以根据需要调整大小

# 假设x是epoch列表，y等是各模型的Time consumption列表
x = range(len(multi_NET_history))  # 假设multi_NET_history是该模型的时间消耗列表

# 绘制散点图
# 绘制散点图
plt.scatter(x, y, marker='o', color='#fcf1f0', label='Q - learning')
plt.scatter(x, y_pro, marker='s', color='#fcccbc', label='dynamic multi - step Q - learning')
plt.scatter(x, step5_Q_history, marker='^', color='#bdb5e1', label='5 - step Q - learning')
plt.scatter(x, step3_Q_history, marker='d', color='#b0d992', label='3 - step Q - learning')
plt.scatter(x, DQN_history_1, marker='*', color='#f9d580', label='DQN')
plt.scatter(x, multi_NET_history, marker='*', color='#99b9e9', s=160, label='our algorithm')

# 绘制平滑连接曲线
plt.plot(x, y, color='#fcf1f0', linestyle='--', alpha=0.7, label='Q - learning')
plt.plot(x, y_pro, color='#fcccbc', linestyle='--', alpha=0.7, label='dynamic multi - step Q - learning')
plt.plot(x, step5_Q_history, color='#bdb5e1', linestyle='--', alpha=0.7, label='5 - step Q - learning')
plt.plot(x, step3_Q_history, color='#b0d992', linestyle='--', alpha=0.7, label='3 - step Q - learning')
plt.plot(x, DQN_history_1, color='#f9d580', linestyle='--', alpha=0.7, label='DQN')
plt.plot(x, multi_NET_history, color='#99b9e9', linestyle='-', linewidth=2, alpha=0.8, label='our algorithm')

plt.xlabel('Epoch to Reach Minimum Time Consumption', fontsize=20)  # 调整标签和字体大小
plt.ylabel('Time Consumption', fontsize=20)
plt.grid(True, linestyle='--', alpha=0.5)
plt.legend(fontsize=16)  # 调整图例字体大小
plt.show()